File: mpikurs_solutions/py/synch_s_ready.py
#synch_s.py
from mpi4py import MPI
import numpy as np

comm = MPI.COMM_WORLD
nproc = comm.Get_size()
myid = comm.Get_rank()

sbuf = np.ones(1)
rbuf = np.empty(1)
t0 = MPI.Wtime() 
# exchange messages between task 0 and all other tasks
# then send messages from task 0 to all other tasks
#   syntax:
#      comm.Send(sbuf,dest=taskid)
#      comm.Recv(rbuf,source=taskid)
for ip in np.arange(1,nproc):
   if myid == 0:
      comm.Send(sbuf,ip)
   if myid == ip:
      comm.Recv(rbuf,0)
for ip in np.arange(1,nproc):
   if myid == ip:
      comm.Send(sbuf,0)
   if myid == 0:
      comm.Recv(rbuf,ip)

dt = MPI.Wtime() - t0
tdmax=comm.reduce(dt,MPI.MAX)


if myid == 0:
   print ("Barriere fuer %i Prozessoren dauert: %e" %(nproc, tdmax))
   print ("Das sind %e Zyklen" % (tdmax*2400.0e+6))

File: mpikurs_solutions/c/synch_s_ready.c
#include<mpi.h>
#include<stdio.h>

int main(argc,argv)
int  argc; char *argv[];
{
   int np, me, msg, ip, i;
   double tm, t0, t1, tbmax;

   MPI_Init(&argc,&argv);
   MPI_Comm_size(MPI_COMM_WORLD,&np);
   MPI_Comm_rank(MPI_COMM_WORLD,&me);

   MPI_Barrier( MPI_COMM_WORLD );

   tm = MPI_Wtime();
   t0 = MPI_Wtime();
   msg = 1;
/*! exchange messages between task 0 and all other tasks
! then send messages from task 0 to all other tasks
!   syntax:
        MPI_Send( &sbuf, 1, MPI_INT, taskid, tagnumber, MPI_COMM_WORLD);
        MPI_Recv( &rbuf, 1, MPI_INT, taskid, tagnumber, MPI_COMM_WORLD,MPI_STATUS_IGNORE);
*/
   for (ip=1;ip<np;ip++){
      if (me == ip) {
        MPI_Send( &msg, 1, MPI_INT, 0, 0, MPI_COMM_WORLD);
      }
      if (me == 0){
        MPI_Recv( &msg, 1, MPI_INT, ip, 0, MPI_COMM_WORLD,MPI_STATUS_IGNORE);
      }
   }
   for (ip=1;ip<np;ip++){
      if (me == 0){
        MPI_Send( &msg, 1, MPI_INT, ip, 0, MPI_COMM_WORLD);
      }
      if (me == ip) {
        MPI_Recv( &msg, 1, MPI_INT, 0, 0, MPI_COMM_WORLD,MPI_STATUS_IGNORE);
      }
   }
   t1 = MPI_Wtime() - t0;
   MPI_Reduce( &t1, &tbmax, 1, MPI_DOUBLE, MPI_MAX, 0,  MPI_COMM_WORLD );

   if (me == 0) {
      printf("Barriere fuer %i Prozessoren dauert : %e\n",np, tbmax);
   }

   MPI_Finalize();
   return 0;
}
